Support cancellation for find_lsn_for_timestamp API

shanyp opened the pull request:
## Problem
See #5900
## Summary of changes
Added cancellation token as param in all relevant code paths and actually used it in the find_lsn_for_timestamp main loop

hlinnaka commented on pageserver/src/bin/pageserver.rs:
```diff
@@ -626,6 +626,7 @@ fn start_pageserver(
+                    cancel,
                 )
                 .instrument(info_span!("metrics_collection"))
                 .await?;
```
Will this throw an error if cancellation happened? In the `tokio::select!` block above, we return `Ok(())` on cancellation. I didn't look closely but ISTM that we should do return the same thing in both cases.

shanyp replied:
44111b3f86d62d458eca33c3eadc81de4711c2a5

hlinnaka commented on pageserver/src/tenant.rs:
```diff
@@ -2569,12 +2570,13 @@ impl Tenant {
         horizon: u64,
         pitr: Duration,
         ctx: &RequestContext,
+        cancel: &CancellationToken,
```
This function does a fair amount of work iterating through each timeline, without checking for cancellation. The cancellation token is only passed to refresh_gc_info_internal(), and not checked after that.

That's not wrong as such, but it means that the function might run a lot longer than you'd think, after cancellation has been requested. The loop does check for `task_mgr::is_shutdown_requested()`, so that would probably be a good place to check for the cancellation token too.

shanyp replied:
37895c21fa8e663b91ba594170654f168fa081c9

hlinnaka commented on pageserver/src/tenant.rs:
```diff
@@ -2569,12 +2570,13 @@ impl Tenant {
         let gc_timelines = self
-            .refresh_gc_info_internal(target_timeline_id, horizon, pitr, ctx)
+            .refresh_gc_info_internal(target_timeline_id, horizon, pitr, ctx, cancel)
             .await?;
```
Do we want this to return an error on cancellation? In the loop below, where we check for `task_mgr::is_shutdown_requested()', we return Ok if requested to shut down. Not sure what the right answer is, but seems inconsistent.

hlinnaka reviewed (approved):
A step in the right direction for sure. A few places are inconsistent on whether you return an error or Ok on cancellation. Would be good to have be consistent on those, but I think this will work even as it is.

shanyp commented:
> A step in the right direction for sure. A few places are inconsistent on whether you return an error or Ok on cancellation. Would be good to have be consistent on those, but I think this will work even as it is.

@hlinnaka thanks, I have addressed your comments , please take another look

shanyp commented on pageserver/src/tenant.rs:
```diff
@@ -2568,14 +2569,30 @@ impl Tenant {
+                {
+                    // Handle cancellation
+                    totals.elapsed = now.elapsed();
+                    return Ok(totals);
```
we do the same below when shutdown is requested, we return ok
